sw/device/silicon_creator/lib/drivers/usb.c

```diff
@@ -0,0 +1,463 @@
+    in_endpoints[i].type = type;
+    in_endpoints[i].max_packet_size = max_packet_size;
+    in_endpoints[i].transfer = (usb_transfer_t){0};
+    in_endpoints[i].handler = handler;
+    in_endpoints[i].user_ctx = user_ctx;
+
+    usbreg_bit(USBDEV_EP_IN_ENABLE_REG_OFFSET, i, true);
+  }
+
+  // Clear stall.
+  usb_ep_stall(ep, false);
+  return kErrorOk;
+}
+
+void fill_fifos(void) {
+  while (!buffer_pool_empty()) {
+    uint32_t status = abs_mmio_read32(kBase + USBDEV_USBSTAT_REG_OFFSET);
+    uint32_t av_setup_depth =
+        bitfield_field32_read(status, USBDEV_USBSTAT_AV_SETUP_DEPTH_FIELD);
+    if (av_setup_depth < 2) {
+      // Supply Available SETUP Buffer
+      uint8_t id = buffer_pool_get();
+      abs_mmio_write32(kBase + USBDEV_AVSETUPBUFFER_REG_OFFSET, id);
+    } else {
+      if (bitfield_bit32_read(status, USBDEV_USBSTAT_AV_OUT_FULL_BIT)) {
+        // The available OUT buffer FIFO is full.
+        break;
+      }
+      // Supply Available OUT Buffer
+      uint8_t id = buffer_pool_get();
+      abs_mmio_write32(kBase + USBDEV_AVOUTBUFFER_REG_OFFSET, id);
+    }
+  }
+}
+
+static bool rx_fifo_empty(void) {
+  uint32_t status = abs_mmio_read32(kBase + USBDEV_USBSTAT_REG_OFFSET);
+  return bitfield_bit32_read(status, USBDEV_USBSTAT_RX_EMPTY_BIT);
+}
+
+static void send_packet(uint8_t ep_index) {
+  usb_ep_info_t *endpoint = in_endpoints + ep_index;
+  size_t chunk = endpoint->max_packet_size < endpoint->transfer.len
+                     ? endpoint->max_packet_size
+                     : endpoint->transfer.len;
+  uint8_t buffer = buffer_pool_get();
+
+  if (chunk < endpoint->max_packet_size) {
+    // If the chunk is shorter than the endpoint size, then we can
+    // clear the ShortIn flag.
+    endpoint->transfer.flags &= ~(uint32_t)kUsbTransactionFlagsShortIn;
+  }
+  copy_to_buffer(buffer, (char *)endpoint->transfer.data, chunk);
+  endpoint->transfer.data += chunk;
+  endpoint->transfer.len -= chunk;
+
+  uint32_t val = 0;
+  val = bitfield_field32_write(val, USBDEV_CONFIGIN_0_BUFFER_0_FIELD, buffer);
+  val = bitfield_field32_write(val, USBDEV_CONFIGIN_0_SIZE_0_FIELD, chunk);
+
+  // Mark the packet as ready for transmission
+  val = bitfield_bit32_write(val, USBDEV_CONFIGIN_0_RDY_0_BIT, true);
+  abs_mmio_write32(
+      kBase + USBDEV_CONFIGIN_0_REG_OFFSET + ep_index * sizeof(uint32_t), val);
+}
+
+rom_error_t usb_ep_transfer(uint8_t ep, void *data, size_t len,
+                            usb_transaction_flags_t flags) {
+  uint8_t i = ep & kEpNumMask;
+  if (i >= USBDEV_PARAM_N_ENDPOINTS) {
+    return kErrorUsbBadEndpointNumber;
+  }
+  usb_ep_info_t *endpoint = ep & kUsbDirIn ? in_endpoints : out_endpoints;
+  endpoint += i;
+
+  if (endpoint->type & kUsbEpTypeControl && len > 0) {
+    // Transfers of more than length zero on a control endpoint require a
+    // zero-length transfer in the opposite direction to finish the transaction.
+    flags |= kUsbTransactionFlagsZlp;
+  }
```

Comment:
Shouldn't the caller be able to handle this, though? Is this here just in case?

Reply:
Yes the caller _could_ handle this, but its here for convenience since any control transfer is going to want this.